File: DesignPatterns/01_GOF23/DP_examples/Rust/01_Creational/Builder/product.rs
// product.rs
// 产品和包装实现

use super::builder::{IPacker, Item};

// 包装器
pub struct Wrapper {
    item_name: String,
}

impl Wrapper {
    pub fn new(item_name: &str) -> Self {
        Self {
            item_name: item_name.to_string(),
        }
    }
}

impl IPacker for Wrapper {
    fn pack(&self, item: &dyn Item) {
        println!("Wrapper a {}", item.get_name());
    }
}

pub struct Bottle {
    item_name: String,
}

impl Bottle {
    pub fn new(item_name: &str) -> Self {
        Self {
            item_name: item_name.to_string(),
        }
    }
}

impl IPacker for Bottle {
    fn pack(&self, item: &dyn Item) {
        println!("Bottle a {}", item.get_name());
    }
}

// Item 实现
pub struct VegetableBurger {
    wrapper: Wrapper,
}

impl VegetableBurger {
    pub fn new() -> Self {
        Self {
            wrapper: Wrapper::new("VegetableBurger"),
        }
    }
}

impl Default for VegetableBurger {
    fn default() -> Self {
        Self::new()
    }
}

impl Item for VegetableBurger {
    fn get_name(&self) -> &str {
        "VegetableBurger"
    }
    
    fn get_price(&self) -> f32 {
        8.5
    }
    
    fn get_packer(&self) -> &dyn IPacker {
        &self.wrapper
    }
}

pub struct ChickenBurger {
    wrapper: Wrapper,
}

impl ChickenBurger {
    pub fn new() -> Self {
        Self {
            wrapper: Wrapper::new("ChickenBurger"),
        }
    }
}

impl Default for ChickenBurger {
    fn default() -> Self {
        Self::new()
    }
}

impl Item for ChickenBurger {
    fn get_name(&self) -> &str {
        "ChickenBurger"
    }
    
    fn get_price(&self) -> f32 {
        15.0
    }
    
    fn get_packer(&self) -> &dyn IPacker {
        &self.wrapper
    }
}

pub struct CokeCola {
    bottle: Bottle,
}

impl CokeCola {
    pub fn new() -> Self {
        Self {
            bottle: Bottle::new("CokeCola"),
        }
    }
}

impl Default for CokeCola {
    fn default() -> Self {
        Self::new()
    }
}

impl Item for CokeCola {
    fn get_name(&self) -> &str {
        "CokeCola"
    }
    
    fn get_price(&self) -> f32 {
        3.5
    }
    
    fn get_packer(&self) -> &dyn IPacker {
        &self.bottle
    }
}

pub struct PepsiCola {
    bottle: Bottle,
}

impl PepsiCola {
    pub fn new() -> Self {
        Self {
            bottle: Bottle::new("PepsiCola"),
        }
    }
}

impl Default for PepsiCola {
    fn default() -> Self {
        Self::new()
    }
}

impl Item for PepsiCola {
    fn get_name(&self) -> &str {
        "PepsiCola"
    }
    
    fn get_price(&self) -> f32 {
        3.0
    }
    
    fn get_packer(&self) -> &dyn IPacker {
        &self.bottle
    }
}

File: DesignPatterns/01_GOF23/DP_examples/C/01_Creational/Builder/Product.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "Product.h"

// 包装器实现
static void wrapper_pack(Burger* item) {
    printf("Wrapper a %s\n", item->Name);
}

static void bottle_pack(ColdDrink* item) {
    printf("Bottle a %s\n", item->Name);
}

// 包装器实例
static IPacker Wrapper_Instance = { wrapper_pack };
static IPacker Bottle_Instance = { bottle_pack };


// 产品实例
Burger VegetableBurger = {
    "VegetableBurger",
    8.5f,
    &Wrapper_Instance
};

Burger ChickenBurger = {
    "ChickenBurger",
    15.0f,
    &Wrapper_Instance
};

ColdDrink CokeCola = {
    "CokeCola",
    3.5f,
    &Bottle_Instance
};

ColdDrink PepsiCola = {
    "PepsiCola",
    3.0f,
    &Bottle_Instance
};


static void Item_destroy(Item* self) {
    if (self) {
        free(self);
    }
}
Item* Item_create(ItemKind kind) {
    Item* item = malloc(sizeof(Item));
    if (item) {
        switch (kind) {
            case VEGETABLE_BURGER:
                memcpy_s(item, sizeof(Burger), &VegetableBurger, sizeof(Burger));
                item->destroy = Item_destroy;
                break;
            case CHICKEN_BURGER:
                memcpy_s(item, sizeof(Burger), &ChickenBurger, sizeof(Burger));
                item->destroy = Item_destroy;
                break;
            case COKE_COLA:
                memcpy_s(item, sizeof(ColdDrink), &CokeCola, sizeof(ColdDrink));
                item->destroy = Item_destroy;   
                break;
            case PEPSI_COLA:
                memcpy_s(item, sizeof(ColdDrink), &PepsiCola, sizeof(ColdDrink));
                item->destroy = Item_destroy;
                break;
        }
    }
    return item;
} 

